### ISSUE TITLE
A Question About ZeRO-Infinity Paper

### ISSUE THREAD

**Issue by vycezhong** [microsoft/DeepSpeed#1137]
In Section 3,  
> The model states are comprised of optimizer states, gradients, and parameters. For mixed precision training with Adam optimizer, the parameters and gradients are stored in FP16 while the optimizer states consist of FP32 momentum, variance, parameters, and gradients. In total, each parameter requires 20 bytes of memory.

Why do you need an FP32 copy of gradients? Is it used for gradient accumulation? 

In the illustration of ZeRO-Offload, I didn't see it.
![image](https://user-images.githubusercontent.com/25879526/120911399-db929c00-c6b9-11eb-817f-5d369407dd16.png)

**Comment by tjruwase** [comment]
@vycezhong, thanks for your question. The optimizer runs in fp32 mode and so requires fp32 gradients.

**Comment by vycezhong** [comment]
Thanks for you explanation. I have another question. 

I notice that the CPU memory bandwidth per GPU drops to 3.0GB/s when using multiple cards. How is this value (3.0GB/s) obtained?
![image](https://user-images.githubusercontent.com/25879526/121275087-74b1f480-c8fe-11eb-9e76-0bd13d3604a0.png)

**Comment by samyam** [comment]
@vycezhong In a 16x V100 DGX-2 box, when all GPUs are simultaneously sending and receiving data to and from CPU memory, we observe that each GPU achieves about 3GB/s for an aggregate bandwidth of about 3*16GB/s. The drop from 12 GB/s to 3GB/s per GPU when going from 1 to 16 GPUs is due to sharing of some of the PCI-E resources. We measured these numbers using micro benchmarks where each GPU sends/receives a few giga bytes of data to/from CPU memory.

**Comment by vycezhong** [comment]
@samyam Thanks. And I found PCIe interconnect topology might explain this as shown in the figure below. Each PCIe link bandwidth is 12GB/s. Along the binary tree from top to bottom, the bandwidth of children should be halved. 

![image](https://user-images.githubusercontent.com/25879526/121306381-587b7b00-c931-11eb-8591-b57c54e80e6a.png)

**Comment by tjruwase** [comment]
@vycezhong, do you have additional questions or can this issue be closed?

**Comment by anijmlt** [comment]
> @vycezhong, thanks for your question. The optimizer runs in fp32 mode and so requires fp32 gradients.

@tjruwase If you need fp32 gradients in the optimizer, are the gradients generated and maintained in fp32 even through their computation on GPU, reduce-scatter phase, offload and storage?
